internal/controller/ops/watchoperation/reconciler.go

```diff
@@ -0,0 +1,233 @@
+		wo.Status.LastSuccessfulTime = &metav1.Time{Time: t}
+	}
+
+	// Count resources being watched. This is best effort. If we hit an
+	// error we just don't update it this time around.
+	ul := &unstructured.UnstructuredList{}
+	ul.SetGroupVersionKind(schema.FromAPIVersionAndKind(wo.Spec.Watch.APIVersion, wo.Spec.Watch.Kind))
+	if err := r.client.List(ctx, ul, client.InNamespace(wo.Spec.Watch.Namespace), client.MatchingLabels(wo.Spec.Watch.MatchLabels)); err == nil {
+		wo.Status.WatchingResources = int64(len(ul.Items))
+	}
+
+	// Garbage collect Operations older than the history limits.
+	for _, op := range lifecycle.MarkGarbage(ptr.Deref(wo.Spec.SuccessfulHistoryLimit, 3), ptr.Deref(wo.Spec.FailedHistoryLimit, 1), ol.Items...) {
+		if err := r.client.Delete(ctx, &op); resource.IgnoreNotFound(err) != nil {
+			log.Debug("Cannot garbage collect Operation", "error", err, "operation", op.GetName())
+			err = errors.Wrapf(err, "cannot garbage collect Operation %q", op.GetName())
+			r.record.Event(wo, event.Warning(reasonGarbageCollect, err))
+			status.MarkConditions(xpv1.ReconcileError(err))
+			_ = r.client.Status().Update(ctx, wo)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Update status with running operations.
+	running := make([]string, 0)
+	for _, op := range ol.Items {
+		if op.GetCondition(v1alpha1.TypeSucceeded).Reason == v1alpha1.ReasonPipelineRunning {
+			running = append(running, op.GetName())
+		}
+	}
+	wo.Status.RunningOperationRefs = lifecycle.RunningOperationRefs(running)
+
+	// Start the Watched controller.
+	wr := watched.NewReconciler(r.engine.GetCached(), wo,
+		watched.WithLogger(r.log.WithValues("controller", WatchedControllerName(wo.GetName()))),
+		watched.WithRecorder(r.record.WithAnnotations("controller", WatchedControllerName(wo.GetName()))))
+
+	ko := r.options.ForControllerRuntime()
+	ko.Reconciler = ratelimiter.NewReconciler(WatchedControllerName(wo.GetName()), errors.WithSilentRequeueOnConflict(wr), r.options.GlobalRateLimiter)
+
+	name := WatchedControllerName(wo.GetName())
+	co := []engine.ControllerOption{engine.WithRuntimeOptions(ko)}
+
+	if err := r.engine.Start(name, co...); err != nil {
+		log.Debug("Cannot start watched resource controller", "error", err)
+		err = errors.Wrap(err, "cannot start watched resource controller")
+		r.record.Event(wo, event.Warning(reasonEstablishWatched, err))
+		status.MarkConditions(v1alpha1.WatchFailed(err.Error()), xpv1.ReconcileError(err))
+		_ = r.client.Status().Update(ctx, wo)
+		return reconcile.Result{}, err
+	}
+
+	// Start watching the specified kind of resource.
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(schema.FromAPIVersionAndKind(wo.Spec.Watch.APIVersion, wo.Spec.Watch.Kind))
+
+	if err := r.engine.StartWatches(ctx, name, engine.WatchFor(u, WatchTypeWatchOperation, NewWatchedResourceHandler(wo))); err != nil {
```

Comment:
i refreshed my memory on https://github.com/crossplane/crossplane/pull/6462 just to remember what was happening in that engine watch error issue, but should be all good in our usage here 😇 

* https://github.com/crossplane/crossplane/pull/6462